packages/astro/src/core/middleware/vite-plugin.ts

```diff
@@ -0,0 +1,138 @@
+import type { Plugin as VitePlugin } from 'vite';
+import { normalizePath } from 'vite';
+import type { Logger } from '../logger/core.js';
+import { getOutputDirectory } from '../../prerender/utils.js';
+import { MIDDLEWARE_PATH_SEGMENT_NAME } from '../constants.js';
+import { addRollupInput } from '../build/add-rollup-input.js';
+import type { BuildInternals } from '../build/internal.js';
+import type { StaticBuildOptions } from '../build/types.js';
+import type { AstroSettings } from '../../@types/astro.js';
+
+export const MIDDLEWARE_MODULE_ID = '@astro-middleware';
+const EMPTY_MIDDLEWARE = '\0empty-middleware';
+
+export function vitePluginMiddleware({
+	logger,
+	settings
+}: {
+	logger: Logger,
+	settings: AstroSettings
+}): VitePlugin {
+	let viteCommand: 'build' | 'serve' = 'serve';
```

Comment:
Since the variant `serve` is never used, maybe this `viteCommand` should be `isCommandBuild: boolean`, and we check if `command === 'build'`